File: src/Rust/killer.rs
use std::collections::HashSet;
use std::collections::HashMap;
use std::process;
use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    
    let f = File::open("/home/sandy/Documents/puzzle.csv").unwrap();
    let f = BufReader::new(f);
	let values: Vec<Vec<usize>> = f.lines()
		.filter_map(
		|l| l.ok().map(
			|s| s.split(",")
			.filter_map(|word| word.parse().ok())
			.collect()))
		.collect();
	let mut parms = HashMap::new();
    
    for v in &values {
		parms.insert( v[v.len()-1], v);
        }
    println!("{:?}", parms);	

    fn same_row(i: usize, j: usize) -> bool {
        return (i / 9) >> 0 == (j / 9) >> 0;
    }

    fn same_col(k: usize, l: usize) -> bool {
        if k > l {
            return (k - l) % 9 == 0;
        } else {
            return (l - k) % 9 == 0;
        }
    }

    fn same_block(m: usize, n: usize) -> bool {
        return (m / 27) >> 0 == (n / 27) >> 0 && ((m % 9) / 3) >> 0 == ((n % 9) / 3) >> 0;
    }

    fn check_row(row_to_check: &mut Vec<usize>, vals: &HashMap<usize,&Vec<usize>>) -> bool {
        let pos: usize;
        if row_to_check.iter().position(|&x| x == 0).is_none() {
            pos = 80;
        } else {
            pos = row_to_check.iter().position(|&s| s == 0).unwrap();
        }
        if vals.contains_key(&(pos-1)) {
            let selected_parm: Vec<usize> = vals.get(&(pos -1)).unwrap().to_vec();
            let sub_parm = &selected_parm[1..selected_parm.len()];
            let total = sub_parm
                .iter()
                .map(|&cell_position| row_to_check[cell_position])
                .fold(0, |total, next| total + next);
            if total != selected_parm[0] {
                return false;
            } else {
                return true;
            }
        } else {
            return true;
        }
    }

    fn recursive_check(in_board: &[usize], vals: &HashMap<usize,&Vec<usize>>) -> bool {
        if in_board.iter().position(|&x| x == 0).is_none() {
            return true;
        } else {
            let i: usize = in_board.iter().position(|&s| s == 0).unwrap();
            let mut excluded_numbers: HashSet<usize> = vec![].into_iter().collect();
            let mut x: usize = 0;
            let mut y: usize = 1;

            while x < i {
                if same_row(i, x) || same_col(i, x) || same_block(i, x) {
                    excluded_numbers.insert(in_board[x]);
                }

                x = x + 1;
            }

            while y < 10 {
                if !excluded_numbers.contains(&y) {
                    let mut new_board: Vec<usize> =
                        [&in_board[..i], &[y], &in_board[(i + 1) ..]].concat();
                    if check_row(&mut new_board, vals) {
                        if i < 80 {
                            recursive_check(&new_board, &vals);
                        } else {
                            println!("resolved board: {:?}", new_board);
                            process::exit(1);
                        }
                    }
                }
                y = y + 1;
            }
            return true;
        }
    }

    let initial_row: [usize; 81] = [0; 81];
    recursive_check(&initial_row, &parms);
}

File: src/TypeScript/killer.ts
const fs = require('fs')
const startTime: Date = new Date()
var values: number[][] = [[]]

try {
  const lines = fs.readFileSync('puzzle.csv', 'utf8').split('\n')
  for (var line of lines) {
    values.push(line.split(',').map(Number)); 
  }
  values.splice(0, 1)

} catch (err) {
  console.error(err)
}

const parms: number[][] = values.sort(
  (a, b) => a[a.length - 1] - b[b.length - 1]
);
const digits: number[] = [1, 2, 3, 4, 5, 6, 7, 8, 9];

let a: number[] = new Array(81).fill(0);
recursiveCheck(a);

function sameRow(i: number, j: number) {
  return (i / 9) >> 0 == (j / 9) >> 0;
}
function sameCol(k: number, l: number) {
  return (k - l) % 9 == 0;
}
function sameBlock(m: number, n: number) {
  return (
    (m / 27) >> 0 == (n / 27) >> 0 && ((m % 9) / 3) >> 0 == ((n % 9) / 3) >> 0
  );
}

function checkRow(aRow: number[]) {
  var pos: number = aRow.indexOf(0);
  let selectedParms: number[][] = parms.filter(
    (parm) => parm[parm.length - 1] == pos - 1
  );
  if (selectedParms.length > 0) {
    let selectedParm: number[] = selectedParms[0];
    let subParm: number[] = selectedParm.slice(1, selectedParm.length);
    let total: number = subParm.map((a) => aRow[a]).reduce((b, c) => b + c, 0);
    if (total != selectedParm[0]) {
      return false;
    } else {
      return true;
    }
  } else {
    return true;
  }
}

function recursiveCheck(inBoard: number[]) {
  let i: number = inBoard.indexOf(0);
  if (i == -1) {
    console.log(inBoard);
    console.log(((new Date()).getTime() - startTime.getTime()) / 1000.0)
    return process.exit(0);
  }

  let excludedNumbers = new Set();
  let x: number = 0;
  let y: number = 1;

  while (x < i) {
    if (sameRow(i, x) || sameCol(i, x) || sameBlock(i, x)) {
      excludedNumbers.add(inBoard[x]);
    }

    x++;
  }

  while (y < 10) {
    if (!excludedNumbers.has(y)) {
      let newBoard: number[] = inBoard
        .slice(0, i)
        .concat([y])
        .concat(inBoard.slice(i, 80));
      if (checkRow(newBoard)) {
        recursiveCheck(newBoard);
      }
    }
    y++;
  }
}
